File: ForexConnectAPIRobot/samples/java/SimpleSamples/TableManagerSamples/CreateOTO/src/ResponseListener.java
package createoto;

import com.fxcore2.IO2GResponseListener;
import com.fxcore2.O2GResponse;
import com.fxcore2.O2GResponseType;

public class ResponseListener implements IO2GResponseListener {

    private boolean mError = false;
    
    // Shows if there was an error during the request processing
    public final boolean hasError() {
        return mError;
    }
    
    // Implementation of IO2GResponseListener interface public method onRequestCompleted
    public void onRequestCompleted(String requestID, O2GResponse response) {
        if ((response.getType() == O2GResponseType.CREATE_ORDER_RESPONSE) ||
            (response.getType() == O2GResponseType.COMMAND_RESPONSE)) {
            System.out.println("Request completed.\nrequestID= " + requestID);
            mError = false;
        } 
    }   

    // Implementation of IO2GResponseListener interface public method onRequestFailed
    public void onRequestFailed(String requestID, String error) {
            System.out.println("Request failed.\nrequestID= " + requestID + "; error= " + error);
            mError = true;  
    }

    // Implementation of IO2GResponseListener interface public method onTablesUpdates
    public void onTablesUpdates(O2GResponse response) {
        
    }   
    
}

File: ForexConnectAPI-1.1.2-Linux-i686/samples/cpp/SimpleSamples/NonTableManagerSamples/RemoveFromGroup/source/ResponseListener.cpp
#include "../../../../../../include/Order2Go2.h"
#include "ResponseListener.h"
#include "stdio.h"

/** Constructor. */
ResponseListener::ResponseListener()
{
    mLastOrdersResponse = NULL;
    mEvent = uni::CreateEvent(0, FALSE, FALSE, 0);
    mRefCount = 1;
}

/** Destructor. */
ResponseListener::~ResponseListener()
{
    if (mLastOrdersResponse != NULL)
        mLastOrdersResponse->release();
}

/** Increase reference counter. */
long ResponseListener::addRef()
{
    return InterlockedIncrement(&mRefCount);
}

/** Decrease reference counter. */
long ResponseListener::release()
{
    long rc = InterlockedDecrement(&mRefCount);
    if (rc == 0)
        delete this;
    return rc;
}

void ResponseListener::onRequestCompleted(const char *requestId, IO2GResponse *response)
{
    if (response->getType() == ::GetOrders)
    {
        mLastOrdersResponse = response;
        mLastOrdersResponse->addRef();
    }
    printf("Request completed.\nrequestID=%s", requestId);
    uni::SetEvent(mEvent);
}

void ResponseListener::onRequestFailed(const char *requestId, const char *error)
{
    printf("Request failed.\nrequestID=%s; error=%s", requestId, error);
    uni::SetEvent(mEvent);
}

void ResponseListener::onTablesUpdates(IO2GResponse *tablesUpdates)
{
}

HANDLE ResponseListener::getEvent() const
{
    return mEvent;
}

/** Gets order response. */
IO2GResponse *ResponseListener::getOrdersResponse()
{
    while (mLastOrdersResponse == NULL)
        uni::WaitForSingleObject(mEvent, 250);

    if (mLastOrdersResponse != NULL)
        mLastOrdersResponse->addRef();
    return mLastOrdersResponse;
}

